tests/test_time_slicing.py

```diff
@@ -0,0 +1,139 @@
+        return 1
+
+    def get_dataarray(self, category, split):
+        if category == "state":
+            values = self._state_data
+        elif category == "forcing":
+            values = self._forcing_data
+        else:
+            raise NotImplementedError(category)
+
+        if self.is_forecast:
+            raise NotImplementedError()
+        else:
+            da = xr.DataArray(
+                values, dims=["time"], coords={"time": self._time_values}
+            )
+            # add `{category}_feature` and `grid_index` dimensions
+
+        da = da.expand_dims("grid_index")
+        da = da.expand_dims(f"{category}_feature")
+
+        return da
+
+    def get_standardization_dataarray(self, category):
+        raise NotImplementedError()
+
+    def get_xy(self, category):
+        raise NotImplementedError()
+
+    def get_vars_units(self, category):
+        raise NotImplementedError()
+
+    def get_vars_names(self, category):
+        raise NotImplementedError()
+
+    def get_vars_long_names(self, category):
+        raise NotImplementedError()
+
+
+ANALYSIS_STATE_VALUES = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+FORCING_VALUES = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
+
+
+@pytest.mark.parametrize(
+    "ar_steps,num_past_forcing_steps,num_future_forcing_steps",
+    [[3, 0, 0], [3, 1, 0], [3, 2, 0], [3, 3, 0]],
+)
+def test_time_slicing_analysis(
+    ar_steps, num_past_forcing_steps, num_future_forcing_steps
+):
+    # state and forcing variables have only on dimension, `time`
+    time_values = np.datetime64("2020-01-01") + np.arange(
+        len(ANALYSIS_STATE_VALUES)
+    )
+    assert len(ANALYSIS_STATE_VALUES) == len(FORCING_VALUES) == len(time_values)
+
+    datastore = SinglePointDummyDatastore(
+        state_data=ANALYSIS_STATE_VALUES,
+        forcing_data=FORCING_VALUES,
+        time_values=time_values,
+        is_forecast=False,
+    )
+
+    dataset = WeatherDataset(
+        datastore=datastore,
+        ar_steps=ar_steps,
+        num_future_forcing_steps=num_past_forcing_steps,
+        num_past_forcing_steps=num_future_forcing_steps,
+        standardize=False,
+    )
+
+    sample = dataset[0]
+
+    init_states, target_states, forcing, _ = [
+        tensor.numpy() for tensor in sample
+    ]
+
+    expected_init_states = [0, 1]
+    if ar_steps == 3:
+        expected_target_states = [2, 3, 4]
+    else:
+        raise NotImplementedError()
+
+    if num_past_forcing_steps == num_future_forcing_steps == 0:
+        expected_forcing_values = [[12], [13], [14]]
+    elif num_past_forcing_steps == 1 and num_future_forcing_steps == 0:
+        expected_forcing_values = [[11, 12], [12, 13], [13, 14]]
+    elif num_past_forcing_steps == 2 and num_future_forcing_steps == 0:
+        expected_forcing_values = [[10, 11, 12], [11, 12, 13], [12, 13, 14]]
+    elif num_past_forcing_steps == 3 and num_future_forcing_steps == 0:
+        raise Exception("No idea what this should be ...")
```

Comment:
>  if num_past_forcing_steps > 2 since that would mean that the forcing window extends back in time beyond the two initial states and so would reduce the total number of samples.

Yes, exactly!

I think this is a nice test to have (would have caught an issue we had before with the forcing windowing, but is fixed now :smile: ). Putting what these values should be, but maybe you want to structure things differently:

```suggestion
        expected_init_states = [1, 2]
        expected_target_states = [3, 4, 5]
        expected_forcing_values = [[10, 11, 12, 13], [11, 12, 13, 14], [12, 13, 14, 15]]
```

Reply:
Ok phew, at least I think I understand now 😅 And I know we are on the same page about what the result should be. However, the tests don't pass :)

With `ar_steps = 3, num_past_forcing_steps = 1, num_future_forcing_steps = 0`:

```python
>       np.testing.assert_equal(forcing[:, 0, :], np.array(expected_forcing_values))

tests/test_time_slicing.py:141:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<built-in function eq>, array([[12., 13.],
       [13., 14.],
       [14., 15.]], dtype=float32), array([[11, 12],
       [12, 13],
       [13, 14]])), kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError:
E           Arrays are not equal
E
E           Mismatched elements: 6 / 6 (100%)
E           Max absolute difference: 1.
E           Max relative difference: 0.09090909
E            x: array([[12., 13.],
E                  [13., 14.],
E                  [14., 15.]], dtype=float32)
E            y: array([[11, 12],
E                  [12, 13],
E                  [13, 14]])
```

Reply:
The issue was just caused by a typo in the test, fixed now :)